header-rewriter/GenCallAsm.cpp

```diff
@@ -216,6 +216,72 @@ static void emit_mixed_wrpkru(AsmWriter &aw, const std::string &pkey0,
+// compartment's stack pointer is stored, and compartment_offset_reg will hold
+// a compartment-specific offset.
+static void emit_load_sp_addr(AsmWriter &aw, const std::string &pkey,
+                              const std::string &sp_addr_reg,
+                              const std::string &compartment_offset_reg) {
+  // Compute compartment offset in compartment_offset_reg
+  emit_load_pkey(aw, pkey, compartment_offset_reg);
+  add_asm_line(aw, llvm::formatv("shl $12, %{0}", compartment_offset_reg));
+  // Compute thread offset in sp_addr_reg, adding compartment offset too
+
+  add_asm_line(aw, llvm::formatv("mov ia2_compacted_tid@GOTTPOFF(%rip), %{0}",
+                                 sp_addr_reg));
+  add_asm_line(aw, llvm::formatv("mov %fs:(%{0}), %{0}", sp_addr_reg));
+
+  // add_asm_line(aw, llvm::formatv("mov thread_id, %{0}", sp_addr_reg));
+
+  add_asm_line(aw, llvm::formatv("lea (%{0}, %{1},8), %{1}",
+                                 compartment_offset_reg, sp_addr_reg));
+  // Add base of ia2_stackptrs
+  add_asm_line(
+      aw, llvm::formatv("add ia2_stackptrs@GOTPCREL(%rip), %{0}", sp_addr_reg));
+}
+
+// Emit code to switch stacks, using ia2_stackptrs as storage for the old and
+// new stack pointer.
+// \p sp_addr_reg and \p compartment_offset_reg are register names (sans %
+// prefix) that will be clobbered.
+static void emit_switch_stacks(AsmWriter &aw, const std::string &old_pkey,
+                               const std::string &new_pkey,
+                               const std::string &sp_addr_reg,
+                               const std::string &compartment_offset_reg) {
+  add_comment_line(
+      aw, llvm::formatv(
+              "Compute location to save old stack pointer (in {0}, using {0})",
+              sp_addr_reg, compartment_offset_reg));
+  emit_load_sp_addr(aw, old_pkey, sp_addr_reg, compartment_offset_reg);
+  add_comment_line(aw, "Write the old stack pointer to memory");
+  add_asm_line(aw, llvm::formatv("movq %rsp, (%{0})", sp_addr_reg));
+
+  // After saving the old sp, we switch stacks by loading the new one
+  add_comment_line(
+      aw, llvm::formatv(
+              "Compute location to load new stack pointer (in {0}, using {1})",
+              sp_addr_reg, compartment_offset_reg));
+  // Subtract old compartment offset so sp_addr_reg = ia2_stackptrs + thread
+  // offset
+  add_asm_line(
+      aw, llvm::formatv("sub %{0}, %{1}", compartment_offset_reg, sp_addr_reg));
+  // Compute new compartment offset in compartment_offset_reg
+  emit_load_pkey(aw, new_pkey, compartment_offset_reg);
+  add_asm_line(aw, llvm::formatv("shl $12, %{0}", compartment_offset_reg));
+  // Add new compartment offset to thread offset
+  add_asm_line(
+      aw, llvm::formatv("add %{0}, %{1}", compartment_offset_reg, sp_addr_reg));
+  add_comment_line(aw, "Read the new stack pointer from memory");
+  add_asm_line(aw, llvm::formatv("movq (%{0}), %rsp", sp_addr_reg));
```

Comment:
Minor nit: you might avoid having to read and then write `rsp` separately if you do something like `rsp += (new_offset - old_offset)`.

Reply:
I don't think we can do that here, because the new and old rsp values have both different bases (different stacks) and different offsets (different location in the stack). We actually never examine the value of rsp at all in this code, just store the old one and load the new one. As such, there's no inherent relation between the old and new values of rsp. We *do* do such a delta optimization on the *location* that the old rsp is saved in memory and where the new rsp is loaded from; that's what's happening around L271.